### PR TITLE
[SPARK-15639][SQL] Try to push down filter at RowGroups level for parquet reader

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

The base class `SpecificParquetRecordReaderBase` used for vectorized parquet reader will try to get pushed-down filters from the given configuration. This pushed-down filters are used for RowGroups-level filtering. However, we don't set up the filters to push down into the configuration. In other words, the filters are not actually pushed down to do RowGroups-level filtering. This patch is to fix this and tries to set up the filters for pushing down to configuration for the reader.
## How was this patch tested?

Existing tests should be passed.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
cc @nongli @liancheng

**Comment by viirya** [comment]
also cc @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 1]
```diff
@@ -344,6 +344,11 @@ private[sql] class ParquetFileFormat
       val hadoopAttemptContext =
         new TaskAttemptContextImpl(broadcastedHadoopConf.value.value, attemptId)
 
+      // Try to push down filters when filter push-down is enabled.
+      // Notice: This push-down is RowGroups level, not individual records.
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Can you provide link to the doc saying it is row group level?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 2]
```diff
@@ -578,62 +583,6 @@ private[sql] object ParquetFileFormat extends Logging {
     }
   }
 
-  /** This closure sets various Parquet configurations at both driver side and executor side. */
-  private[parquet] def initializeLocalJobFunc(
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: reason?

**Comment by yhuai** [comment]
Can you provide a test case that shows the problem? Also, can you provide benchmark results of the performance benefit?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: We longer use this two functions.

**Comment by viirya** [comment]
@yhuai As you can see, this is not to fix a bug/problem. So I think it might be hard to provide a test case for it. I will try to do the benchmark.

**Comment by viirya** [comment]
BTW, I can't see any reason not to add a row-group level filter for parquet.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 1 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Also, does parquet support row group level predicate evaluation?

**Comment by yhuai** [comment]
It is a good idea to add it if parquet supports it (I have an impression that parquet does not support it. But maybe I am wrong). I think having benchmark results is a good practice, so we can avoid it hit any obvious issue.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 2 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Nor `overrideMinSplitSize()` as well. I guess https://issues.apache.org/jira/browse/SPARK-10143 issue would be still happening. I wanted to remove this after verifying it but I could have some time to do so..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: We use `org.apache.parquet.filter2.compat.RowGroupFilter.filterRowGroups` in `SpecificParquetRecordReaderBase` to do filtering.

The implementation of `RowGroupFilter` is at https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/filter2/compat/RowGroupFilter.java.

From [this](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/filter2/compat/RowGroupFilter.java#L91), Looks like it does filtering.

**Comment by viirya** [comment]
@yhuai I've run a simple benchmark as following:

    test("Benchmark for Parquet") {
      val N = 1 << 20

      val benchmark = new Benchmark("Parquet reader", N)
      benchmark.addCase("reading Parquet file", 1) { iter =>
        withParquetTable((0 until N).map(i => (101, i)), "t") {
          sql("SELECT _1 FROM t where t._1 < 100").show()
        }
      }
      benchmark.run()
    }

Before this patch:

    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                        34225 / 34225          0.0       32639.5       1.0X

After this patch:

    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                        31350 / 31350          0.0       29897.6       1.0X

**Comment by viirya** [comment]
ping @yhuai I've addressed the comments. Please take a look again. Thanks!

**Comment by viirya** [comment]
ping @yhuai again

**Comment by viirya** [comment]
cc @rxin Can you also take a look of this? This is staying for a while too. Thanks!

**Comment by viirya** [comment]
cc @cloud-fan too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 1 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: Besides, as we use the metadata in merged schema to figure out if a field is optional (i.e. not in all parquet files) or not to decide to push down a filter regarding it, this info has been ignored in `FileSourceStrategy` now. Without the fixing in this change, the push-down row-group level filtering will be failed due to not existing field in parquet file.

**Comment by viirya** [comment]
ping @yhuai @rxin @cloud-fan

**Comment by rxin** [comment]
Is this a bug fix or performance fix? Sorry I don't really understand after reading your description.

**Comment by viirya** [comment]
It is not really a bug fix because without this filtering push-down, the thing still works. This should be a performance fix. I should modify the description.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
The description is updated.

**Comment by yhuai** [comment]
@viirya I took a look at parquet's code. Seems parquet only evaluate row group level filters when generating splits (https://github.com/apache/parquet-mr/blob/apache-parquet-1.7.0/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java#L673). With FileSourceStrategy in Spark, I am not sure we will actually evaluate filter unneeded row groups as expected. Can you take a look? Also, it will be great if you can have a test to make sure that we actually can skip unneeded row groups. This test can be created as follows.

1. We first write a parquet file containing multiple row groups. Also, let's that there is a column `c` and those row groups have disjoint ranges of `c`'s values.
2. We write a query having a filter on `c` and we make sure that this query only need a subset of row groups.
3. We verify that we only create splits for the needed row groups.

**Comment by viirya** [comment]
@yhuai Parquet also does this filtering at ParquetRecordReader (https://github.com/apache/parquet-mr/blob/4b1ff8f4b9dfa0ccb064ef286cf2953bfb2c492d/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetRecordReader.java#L178) and ParquetReader(https://github.com/apache/parquet-mr/blob/4b1ff8f4b9dfa0ccb064ef286cf2953bfb2c492d/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetReader.java#L145).

In Spark, we also did this at SpecificParquetRecordReaderBase (https://github.com/apache/spark/blob/f958c1c3e292aba98d283637606890f353a9836c/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java#L103).

I've manually tested it. But it should be good to have a formal test case for it as you said. I will try to add it later, maybe when I come back to work few days later...

**Comment by viirya** [comment]
@yhuai Your step 3 may not work. We are going to filter the row groups for each parquet file to read in `VectorizedParquetRecordReader`. I think we don't do anything regarding creating splits?

**Comment by liancheng** [comment]
@yhuai We used to support row group level filter push-down before refactoring `HadoopFsRelation` into `FileFormat`, but lost it (by accident I guess) after the refactoring. So now we only have row group level filtering when the vectorized reader is not used, [see here][1].

And yes, both `ParquetInputFormat` and `ParquetRecordReader` do row group level filtering.

This LGTM. Thanks for fixing it! Merging to master and 2.0.

[1]: https://github.com/apache/spark/blob/54f758b5fc60ecb0da6b191939a72ef5829be38c/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala#L371-L378

**Comment by rxin** [comment]
I just talked to @liancheng offline. I don't think we should've merged this until we have verified there is no performance regression, and we definitely shouldn't have merged this in 2.0.

@liancheng can you revert this from both master and branch-2.0?

@viirya can you run some parquet scan benchmark and make sure this does not result in perf regression?

**Comment by rxin** [comment]
To be more clear, please write a proper benchmark that reads data when filter push down is not useful to compare whether this regress performance for the non-push-down case. Also make sure the benchmark does not include the time it takes to write the parquet data.

**Comment by rxin** [comment]
And once we have more data, it might make sense to merge this in 2.0!

**Comment by viirya** [comment]
@rxin One thing needs to be explain is, because we just have one configuration to control filter push down, it affects row-based filter push down and this row-group filter push down.

The benchmark I posted above is running it against this patch and master branch individually. Of course it includes the time to write the parquet data, I will change it. I want to confirm if this kind of benchmark is enough?

**Comment by liancheng** [comment]
Reverted from master and branch-2.0.

@viirya For the benchmark, there are two things:

1. The benchmark also counts Parquet file writing into it, so the real number should be much better than the posted one.
2. We should also benchmark for cases where no filters are pushed down to verify that this patch doesn't affect normal code path.

**Comment by viirya** [comment]
@liancheng Got it.

**Comment by viirya** [comment]
@liancheng 

I rerun the benchmark that excludes the time of writing Parquet file:

    test("Benchmark for Parquet") {
      val N = 1 << 50
        withParquetTable((0 until N).map(i => (101, i)), "t") {
          val benchmark = new Benchmark("Parquet reader", N)
          benchmark.addCase("reading Parquet file", 10) { iter =>
            sql("SELECT _1 FROM t where t._1 < 100").collect()
          }
          benchmark.run()
      }
    }

`withParquetTable` in default will run tests for vectorized reader non-vectorized readers. I only let it run vectorized reader.

After this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17 on Linux 3.13.0-57-generic
    Westmere E56xx/L56xx/X56xx (Nehalem-C)
    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                            76 /   88          3.4         291.0       1.0X

Before this patch:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17 on Linux 3.13.0-57-generic
    Westmere E56xx/L56xx/X56xx (Nehalem-C)
    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                            81 /   91          3.2         310.2       1.0X

Next, I run the benchmark for non-pushdown case using the same benchmark code but with disabled pushdown configuration.

After this patch:

    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                            80 /   95          3.3         306.5       1.0X

Before this patch:

    Parquet reader:                          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    reading Parquet file                            80 /  103          3.3         306.7       1.0X

For non-pushdown case, from the results, I think this patch doesn't affect normal code path.

**Comment by yhuai** [comment]
Can you add results showing that there are skipped row groups with this change (and before this patch all row groups are loaded)?

For those results, let's also put them in the description of the new PR.

**Comment by viirya** [comment]
@yhuai ok. Do you mean I need to create a new PR for this?

**Comment by yhuai** [comment]
Yea. Since this one was closed by asfgit, I am not sure you can reopen it.





On Wed, Jun 15, 2016 at 7:39 PM -0700, "Liang-Chi Hsieh" <notifications@github.com> wrote:












@yhuai ok. Do you mean I need to create a new PR for this?

**Comment by liancheng** [comment]
@viirya One problem in your new benchmark code is that `1 << 50` is actually very small since it's an `Int`:

```
scala> 1 << 50
res0: Int = 262144
```

Anyway, `1 << 50`, which is 1PB, might be too large a value for such a microbenchmark :)

So the generated Parquet file probably only contains a single row group, I guess that's why the numbers are quite close no matter you enable row group filter push-down or not.

**Comment by viirya** [comment]
@liancheng Thanks! I didn't notice that. I will rerun the benchmark. I've re-submitted this PR at #13701.

**Pull request closed without merging**